server/monitor-types/tls.js

```diff
@@ -0,0 +1,266 @@
+            socket.resume();
+        });
+    }
+
+    /**
+     * Reads available data from the given socket if it starts with a given prefix.
+     * @param {AbortController} aborter  Abort controller used to abort the read
+     * @param {*}               socket   net.Socket or tls.TLSSocket instance to use
+     * @param {string}          expected Prefix the response is expected to start with
+     * @returns {Promise<string>} Data read from the socket or rejected promise if the response does
+     *                            not start with the prefix
+     */
+    async expectDataStartsWith(aborter, socket, expected) {
+        const data = await this.readData(aborter, socket);
+        this.debug_log(`response data: '${data}', expected: '${expected}'…`);
+        if (!data.startsWith(expected)) {
+            throw new Error(`Unexpected response. Data: '${data}', Expected: '${expected}'…`);
+        }
+    }
+
+    /**
+     * Performs STARTTLS on the given socket.
+     * @param {AbortController} aborter Abort controller used to abort the STARTTLS process
+     * @param {*}               socket  net.Socket or tls.TLSSocket instance to use
+     * @returns {Promise<void>} Rejected promise if the STARTTLS process failed
+     */
+    async startTls(aborter, socket) {
+        this.debug_log("waiting for prompt");
+        await this.expectDataStartsWith(aborter, socket, "220 ");
+        this.debug_log("sending STARTTLS");
+        socket.write("STARTTLS\n");
+        this.debug_log("waiting for ready-to-TLS");
+        await this.expectDataStartsWith(aborter, socket, "220 ");
+    }
+
+    /**
+     * Upgrades an unencrypted socket to a TLS socket using STARTTLS.
+     * @param {*}               socket  net.Socket representing the unencrypted connection
+     * @returns {Promise<TLSSocket>} tls.TLSSocket instance representing the upgraded TLS connection
+     *                               or rejected promise if the STARTTLS process failed
+     */
+    upgradeConnection(socket) {
+        return new Promise((resolve, reject) => {
+            const onSecureConnect = () => {
+                const cipher = tlsSocket.getCipher();
+                this.debug_log(`connected: authorized: ${tlsSocket.authorized}, cipher: ${cipher ? cipher.name : "<none>"}`);
+                resolve(tlsSocket);
+            };
+
+            const onError = (error) => {
+                reject(error);
+            };
+
+            const tlsSocket = tls.connect({ socket: socket });
+
+            tlsSocket.on("secureConnect", onSecureConnect);
+            tlsSocket.on("error", onError);
+        });
+    }
+
+    /**
+     * Logs a message.
+     * @param {*} msg Message
+     * @returns {void}
+     */
+    log(msg) {
+        log.debug(this.name, msg);
+    }
+
+    /**
+     * Logs a debug message (disabled by default).
+     * @param {*} msg Message
+     * @returns {void}
+     */
+    debug_log(msg) {
+        //log.debug(this.name, msg);
+    }
```

Comment:
That was on purpose since I extracted the STARTTLS functionality from a separate test bed where I'm logging to the console. I can inline the `log()` method but I'd like to leave the no-op `debug_log()` in place if you don't mind. It might come in handy for debugging down the road. (If you have a better idea for not spamming the main log while leaving easy-to-enable debug statements in place, please let me know!)

Reply:
`log.debug` is only printed in development or when the log level is lowered

Reply:
> `log.debug` is only printed in development or when the log level is lowered.

Ah, good to know. I've changed all of them to `log.debug(…)`.

> => `this.log` calling `log.debug` is a bug ^^

It's gone now but I'm still curious where the bug is. (JS isn't my day job either. :grin:) The name lookup rules seem to make sure that the `log` binds to the global object rather than the function and it appears to run just fine.

Reply:
If you want to log at the info level, `log.info` is the correct function to do so. Should have been more precise above.